stdlib/public/core/MutableCollection.swift

```diff
@@ -398,6 +398,29 @@ extension MutableCollection where Self: RandomAccessCollection {
   }
 }
 
+extension MutableCollection {
+  /// Exchange the values at indices `i` and `j`.
+  ///
+  /// Has no effect when `i` and `j` are equal.
+  @_inlineable
+  public mutating func swapAt(_ i: Index, _ j: Index) {
+    guard i != j else { return }
+    
+    // Semantically equivalent to (i, j) = (j, i).
+    // Microoptimized to avoid retain/release traffic.
+    let p1 = Builtin.addressof(&self[i])
+    let p2 = Builtin.addressof(&self[j])
```

Comment:
I would recommend still using the global swap when possible.  You can't do that in a generic implementation of MutableCollection, but you can in a concrete implementation on Array (but it seems you haven't made this a protocol requirement?), and you can in many of the individual places you're doing it.  Recall that we aren't actually eliminating the global swap.